src/Provider/AbstractProvider.php

```diff
@@ -370,20 +360,60 @@ protected function getAccessTokenQuery(array $params)
     }
 
     /**
-     * Checks that a provided grant is valid, or attempts to produce one if the
-     * provided grant is a string.
+     * Returns the full URL to use when requesting an access token.
      *
-     * @param mixed $grant
-     * @return AbstractGrant
+     * @param array $params Query parameters
      */
-    protected function verifyGrant($grant)
+    protected function getAccessTokenUrl(array $params)
     {
-        if (is_string($grant)) {
-            return $this->grantFactory->getGrant($grant);
+        $url = $this->getBaseAccessTokenUrl();
+
+        if ($this->getAccessTokenMethod() === 'GET') {
+            $query = $this->getAccessTokenQuery($params);
+            return $this->appendQuery($url, $query);
         }
 
-        $this->grantFactory->checkGrant($grant);
-        return $grant;
+        return $url;
+    }
+
+    /**
+     * Returns headers that are only used when requesting an access token.
+     *
+     * @return array
+     */
+    protected function getAccessTokenHeaders()
+    {
+        return $this->getDefaultHeaders();
```

Comment:
I think this would be better solved by having `getRequest` also call `getHeaders`. Also, I noticed another thing... any time `getHeaders` is called by `get*Request`, it should be using `array_replace` instead of directly setting `$options['headers']` otherwise user-defined headers will be lost:

```php
$options = array_replace_recursive($options, [
    'headers' => $this->getHeaders(...),
]);
```